shared/src/balancer/swap/weighted_math.rs

```diff
@@ -0,0 +1,153 @@
+//! Module emulating some of the functions in the Balancer WeightedMath.sol
+//! smart contract. The original contract code can be found at:
+//! https://github.com/balancer-labs/balancer-v2-monorepo/blob/6c9e24e22d0c46cca6dd15861d3d33da61a60b98/pkg/core/contracts/pools/weighted/WeightedMath.sol
+
+#![allow(dead_code)]
+
+use super::error::Error;
+use super::fixed_point::Bfp;
+use ethcontract::U256;
+use lazy_static::lazy_static;
+
+// https://github.com/balancer-labs/balancer-v2-monorepo/blob/6c9e24e22d0c46cca6dd15861d3d33da61a60b98/pkg/core/contracts/pools/weighted/WeightedMath.sol#L36-L37
+lazy_static! {
+    static ref MAX_IN_RATIO: Bfp =
+        Bfp::from_wei(U256::exp10(17).checked_mul(3_u32.into()).unwrap());
+    static ref MAX_OUT_RATIO: Bfp =
+        Bfp::from_wei(U256::exp10(17).checked_mul(3_u32.into()).unwrap());
+}
+
+// https://github.com/balancer-labs/balancer-v2-monorepo/blob/6c9e24e22d0c46cca6dd15861d3d33da61a60b98/pkg/core/contracts/pools/weighted/WeightedMath.sol#L69-L100
+pub fn calc_out_given_in(
+    balance_in: Bfp,
+    weight_in: Bfp,
+    balance_out: Bfp,
+    weight_out: Bfp,
+    amount_in: Bfp,
+) -> Result<Bfp, Error> {
```

Comment:
While this may not be necessary, I see there is nothing binding this method to a particular or any WeightedPool. While the method is internal to the balancer module, it still could be defined explicitly for a weighted pool. This might allow us to use attributes from the pool instance like (like weight_in/out, balance_in/out) by just providing the in/out tokens and amount_in.

Just an idea though as it would look a lot more like 

http://github.com/gnosis/gp-v2-services/blob/7c00a326640901ccdbf3414f50d7a13024adf2e5/shared/src/baseline_solver.rs#L11-L15

However, as you mentioned, we would use these to implement BaselineSolvable so I suppose it was your intention to do this next while making use of these emulated contract functions.

Reply:
I think the idea is to use this in the `BaseLineSolvable` implementation for a weighted pool. So something like:
```rust
impl BaseLineSolvable for WeighedPool {
  fn get_amount_out(&self, out_token: H160, in_amount: U256, in_token: H160) -> Option<U256> {
    let reserve_out = self.reserives.get(&out_token)?;
    let reserve_in = self.reserves.get(&in_token)?;
    calc_out_given_in(reserve_in.balance, reserve_in.weight, resrve_out.balance, reserve_out.weight, in_amount).ok()
  }
}
```

This way the method stays nice and easy to test (and check against the Balancer SC implementation) while allowing it to eventually be tied to a weighted pool (which has some other unrelated logic associated with it - such as getting the reserves for the specified tokens).

Reply:
Everything that Nick says is right.
I want to be as close as possible to the smart contract to make it easy to verify, since the actual implementation of `BaselineSolvable` must be different because the interface is different.